TYP: preserve type for split functions via structural Protocol

janosh opened the pull request:
add overloads for `array_split`, `split`, `hsplit`, `vsplit`, and `dsplit` that preserve the input type when the input implements `swapaxes` and `__getitem__`.

this uses structural typing (Protocol) based on the methods actually used by these split functions internally - not `__array_function__`. the Protocol is:

```python
@type_check_only
class _SupportsSplitOps(Protocol):
    def swapaxes(self, axis1: Any, axis2: Any, copy: Any = ...) -> Any: ...
    def __getitem__(self, key: Any, /) -> Any: ...
```

objects like `pandas.DataFrame` that implement these methods (and return `Self` from them) will have their type preserved through split operations.

### example

```python
import numpy as np
import pandas as pd

df = pd.DataFrame({"a": [1, 2, 3, 4]})
splits = np.array_split(df, 2)
splits[0].columns  # before: error (ndarray has no attr "columns"), after: works
```

### why this works

the split functions use `swapaxes` and slicing (`__getitem__`) internally. for objects like DataFrame that return their own type from these operations, the split result preserves the type. this is pure duck-typing, not `__array_function__` dispatch.

closes #23005

janosh commented:
thanks for the feedback! i updated the PR title and description - they were outdated.

to clarify: this PR does **not** use `__array_function__` at all. the approach is purely structural typing via a Protocol:

```python
class _SupportsSplitOps(Protocol):
    def swapaxes(self, axis1: Any, axis2: Any, copy: Any = ...) -> Any: ...
    def __getitem__(self, key: Any, /) -> Any: ...
```

the split functions internally use `swapaxes` and slicing. for objects like `DataFrame` that implement these methods and return `Self`, the type is preserved. this is just duck-typing - no `__array_function__` dispatch involved.

the Protocol matches what the implementation actually does, so it should be type-safe for any object that properly implements these methods.

shoyer commented:
OK, interesting. I'm still hesitant to recommend relying on this behavior, but it's been around in NumPy long enough that it's probaby not going to change.

The real challenge is that strictly following the implementation, we don't know what the return type should be -- it could be an entirely different type, principle.

I would still suggest that anybody concerned with type safety should avoid these implicitly overloaded NumPy APIs. In retrospect, they were clearly a design mistake.

jorenham reviewed (changes requested):
Thanks for this PR. 

Overall this looks fine to me. The shapes are retained, so even if the input arrays include static shape-typing information, this will still be sound.
One issue with this protocol is that it isn't complete. If you look at the `array_split` implementation, for example, you'll see 

```py
try:
    Ntotal = ary.shape[axis]
except AttributeError:
    Ntotal = len(ary)
```

But as it currently stands, there are objects assignable to `_SupportsSplitOps` that have neither a `shape` nor `__len__`. Please take a closer look at each implementation to see what other methods/attributes the `Protocol` should have.

Recently, NumPy dropped support for Python 3.11, so we're using PEP 695 syntax now, in favor of `TypeVar`. You'll see ruff complaining about this if you run `ruff check`.

This needs tests to verify that it is working as intended. We don't have Pandas available in our test environment, so you'll have to ducktype something (that'll also work at runtime) instead. You can find them under `typing/tests/data`

janosh committed:
TYP: preserve type for split functions via structural Protocol

Add overloads for array_split, split, hsplit, vsplit, and dsplit that
preserve the input type when the input implements _SupportsSplitOps.

The Protocol requires:
- shape: tuple[int, ...] - for ary.shape[axis] access
- ndim: int - for dimensional checks in hsplit/vsplit/dsplit
- swapaxes(axis1, axis2) -> Self - for axis manipulation
- __getitem__(key) -> Self - for slicing

Uses PEP 695 syntax and adds tests with a duck-typed SplitableArray
class that implements the protocol.

Closes #23005

jorenham commented on numpy/lib/_shape_base_impl.pyi:
```diff
@@ -57,6 +65,19 @@ class _SupportsArrayWrap(Protocol):
+    def shape(self) -> tuple[int, ...]: ...
+    @property
+    def ndim(self) -> int: ...
+    def swapaxes(self, axis1: SupportsIndex, axis2: SupportsIndex, /) -> Self: ...
```
I'm not sure if this would cause any issues for pandas, but the `axis{1,2}` are in contravariant (input) positions, so it would be more general to annotate them as `: int`.

janosh committed:
use int instead of SupportsIndex for swapaxes params

As suggested in review: axis parameters are in contravariant position,
so using int is more general and allows implementations that accept
int or broader types.

Also: clean up comment, fix import formatting.

janosh commented:
yes, i tested the changes here on https://github.com/janosh/matbench-discovery which uses `np.array_split` on `pd.DataFrame` a lot

jorenham commented:
Thanks @janosh; great use of structural typing :ok_hand:

jorenham commented:
Backporting this will be tricky because of the PEP 695 syntax unfortunately
